lib/algos/list/delimiterVersions.js

```diff
@@ -1,106 +1,165 @@
-        this.latestVersions = undefined; // final list of the latest versions
-        this._latestVersions = latestVersions; // reserved for caching
+    constructor(parameters) {
+        super(parameters);
+        // specific to version listing
+        this.keyMarker = parameters.keyMarker;
+        this.versionIdMarker = parameters.versionIdMarker;
+        // internal state
+        this.masterKey = undefined;
+        this.masterVersionId = undefined;
+        // listing results
+        this.NextMarker = parameters.keyMarker;
+        this.NextVersionIdMarker = undefined;
+    }
+
+    genMDParams() {
+        const params = {};
+        if (this.parameters.prefix) {
+            params.gte = this.parameters.prefix;
+            params.lt = inc(this.parameters.prefix);
+        }
+        if (this.parameters.keyMarker) {
+            if (params.gte && params.gte > this.parameters.keyMarker) {
+                return params;
+            }
+            delete params.gte;
+            if (this.parameters.versionIdMarker) {
+                // versionIdMarker should always come with keyMarker
+                // but may not be the other way around
+                params.gt = formatVersionKey(this.parameters.keyMarker,
+                        this.parameters.versionIdMarker);
+            } else {
+                params.gt = inc(this.parameters.keyMarker + VID_SEP);
+            }
+        }
+        return params;
     }
 
     /**
-     * Overriding the base function to not process the metadata entry here,
-     * leaving the job of extracting object's attributes to S3.
-     *
-     * @param {string} key - key of the entry
-     * @param {string} value - value of the entry
-     * @return {undefined}
+     *  Add a (key, versionId, value) tuple to the listing.
+     *  Set the NextMarker to the current key
+     *  Increment the keys counter
+     *  @param {object} obj             - the entry to add to the listing result
+     *  @param {String} obj.key         - The key to add
+     *  @param {String} obj.versionId   - versionId
+     *  @param {String} obj.value       - The value of the key
+     *  @return {Boolean} - indicates if iteration should continue
      */
-    addContents(key, value) {
-        const components =
-            VSUtils.getObjectNameAndVersionIdFromVersionKey(key);
-        const objectName = components.objectName;
-        const versionId = components.versionId;
-        this.Contents.push({
-            key: objectName,
-            value,
-        });
-        this.NextMarker = objectName;
-        this.NextVersionMarker = versionId;
-        ++this.keys;
-        // only include the latest versions of the keys in the resulting list
-        // this is not actually used now, it's reserved for caching in future
-        if (this._latestVersions) {
-            this.latestVersions[objectName] = this._latestVersions[objectName];
+    addContents(obj) {
```

Comment:
But because this is a short function and we would need to check the returned value from the base implementation to add `this.NextVersionIdMarker = obj.versionId;` so I think it's clearer to do it this way while being as concise.